Sources/BSON/BSONDocumentIterator.swift

```diff
@@ -43,31 +55,39 @@ public struct BSONDocumentIterator: IteratorProtocol {
         }
 
         guard let type = BSONType(rawValue: typeByte), type != .invalid else {
-            fatalError("BSONDocumentIterator Failed: Invalid type, \(typeByte)")
+            throw BSONError.InternalError(message: "BSONDocumentIterator Failed: Invalid type, \(typeByte)")
         }
 
-        do {
-            let key = try self.buffer.readCString()
-            guard let bson = try BSON.allBSONTypes[type]?.read(from: &buffer) else {
-                throw BSONError.InternalError(message: "Cannot read unknown type: \(type)")
-            }
-            return (key: key, value: bson)
-        } catch {
-            fatalError("BSONDocumentIterator.next() failed: \(error)")
+        let key = try self.buffer.readCString()
+        guard let bson = try BSON.allBSONTypes[type]?.read(from: &buffer) else {
+            throw BSONError.InternalError(message: "Cannot read unknown type: \(type)")
         }
+        return (key: key, value: bson)
     }
 
     /// Finds the key in the underlying buffer, and returns the [startIndex, endIndex) containing the corresponding
     /// element.
-    internal func findByteRange(for key: String) -> (startIndex: Int, endIndex: Int) {
-        fatalError("Unimplemented")
+    internal func findByteRange(for searchKey: String) -> (startIndex: Int, endIndex: Int)? {
+        var iter = BSONDocumentIterator(over: self.buffer)
```

Comment:
Since the buffer has to be mutated to track the changes in the readerIndex and currently the iterator's buffer is private I think its fine to leave in here. I've updated to use `self`